### DISCUSSION TITLE
关于 zig async

### DISCUSSION THREAD

**Discussion by kscooo** [zigcc/forum#31] [新手提问]
看到官方文档 async/await 一些简介，没有搜到更多的资料。

我的疑惑是：
官方实现是无栈的实现，那么像 golang goroutine/rust tokio 这样有栈且可以公平调度到多核 runtime 这类实现是由社区实现？
官方所宣传的好处可挂起的无堆栈协程这种实现有啥实质性的好处，没有污染性？但是我个人理解在调度在多核这类公平性上（golang gmp）应该是有欠缺的，请大佬们赐教。

**Comment by jiacai2050** [comment]
问题有些大，关于哪种方式更公平，感觉更多要看使用场景，而且公平的定位对每个人来说都可能不一样，所以还是建议就事论事，举个实际的例子会更利于后面的讨论。

**Reply by kscooo** [reply]
fine，我的问题就是 zig 官方默认的实现本质是协作式，golang/rust 这种有抢占时调度（golang goroutine、tokio）的实现，未来这类实现是否是社区来做？
zig 对 async/await 创新有什么优点。

**Reply by jiacai2050** [reply]
> 未来这类实现是否是社区来做？

这个只能去官方 issue 问问 andrew 了。

> zig 对 async/await 创新有什么优点。

1. 消除 color 的问题，可以仅通过 `pub const io_mode = .evented;` 来切换运行状态
2. Zig 的异步可以在所有平台上运行 https://github.com/ziglang/zig/issues/727#issuecomment-408127967

目前我还没什么时候实战经验，等有了再来补充。

**Comment by jiacai2050** [comment]
看了下[这个](https://github.com/ziglang/zig/issues/6025#issuecomment-1743714049)，稍微总结下，stage2 的 async 现在是卡在两点上：
1. async 函数挂起时，会保存所有上下文，但是在递归函数里，容易 oom
2. 无法推倒出 函数指针 是不是 async fn 的，async fn 与普通 fn 的调用方式是不一样的

**Reply by jiacai2050** [reply]
https://github.com/ziglang/zig/issues/16270#issuecomment-1892908028

这里有了个后续，没想到 LLVM 这么弱 🤪

**Reply by jiacai2050** [reply]
https://github.com/rust-lang/rust/pull/117703

原来 rust 也是在最近在支持的递归调用。

**Comment by jinzhongjia** [comment]
go的goroutine实际上的模型是GMP，M个系统级线程，P个内核handle，G则是goroutine，这个解释是普遍的一个说法

更确切的将，你可以理解为是go官方实现的一个抢占式的用户协程，普通的协程库设计的原则是协程之间互相谦让，而go这边则不需要，由go的调度器实现直接抢占。

并且go的调度器会直接帮开发者实现这个过程，具体说起来可能比较复杂，可以看这篇文章：https://draveness.me/golang/docs/part3-runtime/ch06-concurrency/golang-goroutine/

**Comment by jinzhongjia** [comment]
关于目前zig的async特性，别抱太大期望，按照目前他们的规划，这个是很靠后的工作

**Comment by jiacai2050** [comment]
https://github.com/ziglang/zig/wiki/FAQ#what-is-the-status-of-async-in-zig

最新的动态

**Reply by jinzhongjia** [reply]
实际上并没有明确说是否会将async添加回来，老实说，如果难度过大我怀疑安德鲁会把它砍掉（完美主义者，宁可没有，也不要一个不好用的东西）

**Reply by jiacai2050** [reply]
不要心急，目前阶段了解进展更重要，rust 也是 1.0 后很久才有的 async 的支持。
